File: 백준/Silver/1697. 숨바꼭질/숨바꼭질.py
from collections import deque

N, K = map(int, input().split())
MAX = 100001
dist = [0] * MAX

def bfs(start):
    queue = deque()
    queue.append(start)

    while queue:
        x = queue.popleft()
        if x == K:
            return dist[x]

        for nx in (x - 1, x + 1, x * 2):
            if 0 <= nx < MAX and not dist[nx]:
                dist[nx] = dist[x] + 1
                queue.append(nx)

print(bfs(N))

File: 백준/Silver/1697. 숨바꼭질/숨바꼭질.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.StringTokenizer;
import java.util.Queue;
import java.util.LinkedList;

public class Main {
    static int result;
    static int N, K;
    static int[] dist;
    static int MAX = 100001;

    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());

        N = Integer.parseInt(st.nextToken());
        K = Integer.parseInt(st.nextToken());

        dist = new int[MAX];

        for (int i = 0; i < MAX; i++) {
            dist[i] = -1;
        }

        result = bfs(N, K);

        System.out.println(result);

    }

    public static int bfs(int start, int target) {
        Queue<Integer> q = new LinkedList<>();
        q.add(start);
        dist[start] = 0;

        while(!q.isEmpty()) {
            int x = q.poll();

            if (x == target) {
                return dist[x];
            }

            int[] moves = {x - 1, x + 1, x * 2};

            for (int nx : moves) {
                if (nx >= 0 && nx < MAX && dist[nx] == -1) {
                    dist[nx] = dist[x] + 1;
                    q.add(nx);
                }
            }
        }
        return -1;
    }
}